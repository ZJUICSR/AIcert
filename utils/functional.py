#!/usr/bin/env python
# -*- coding:utf-8 -*-

__author__ = 'homeway'
__copyright__ = 'Copyright © 2021/03/22, ZJUICSR'

import datetime
import os

import redis
from flask_caching import Cache

from function.task import Task


def init(app):
    app.cache = Cache()
    app.cache.init_app(app)

    target_folder = [app.config["LOG_FOLDER"], app.config["OUTPUT_FOLDER"]]
    for folder in target_folder:
        if not os.path.exists(folder):
            try:
                os.makedirs(folder)
            except Exception as e:
                print(f"Write `{folder}` error for read-only file system!")

    try:
        _redis = redis.Redis(host=app.config["REDIS_HOST"], port=app.config["REDIS_PORT"], db=app.config["REDIS_DB"],
                             password=app.config["REDIS_PASS"])
        app.redis = _redis
    except:
        raise ConnectionError("Can not connect to redis server!")


def decode(redis_dict):
    result = {}
    for k in redis_dict.keys():
        v = redis_dict[k]
        result[k.decode("utf-8")] = v.decode("utf-8")
    return result

def init_utils(app):
    app.task = Task(app)


def clean_ex_img():
    path = "web/static/images/result"
    file_list = os.listdir(path)
    for file in file_list:
        file_path = os.path.join(path,file)
        if os.path.exists(file_path):
            os.remove(file_path)

def save_ex_img(img, img_name):
    path = "web/static/images/result/"
    statci_path = "static/images/result/"

    curr_time = datetime.datetime.now()
    current_time = datetime.datetime.strftime(curr_time, '%Y-%m-%d-%H：%M：%S') #需要使用中文引号
    file_name = path + current_time + "_" + img_name
    f = statci_path + current_time + "_" + img_name

    img.save(file_name)
    return f

def save_adv_img(img, img_name):
    path = "web/static/images/adv_images/"
    statci_path = "static/images/adv_images/"

    curr_time = datetime.datetime.now()
    current_time = datetime.datetime.strftime(curr_time, '%Y-%m-%d-%H：%M：%S') #需要使用中文引号
    file_name = path + current_time + "_" + img_name
    f = statci_path + current_time + "_" + img_name

    img.save(file_name)
    return f