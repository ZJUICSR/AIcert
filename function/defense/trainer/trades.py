import torch
import torch.nn as nn
from torch.autograd import Variable
import torch.optim as optim
import torch.nn.functional as F
import numpy as np

def trades_loss(model,
                x_natural,
                y,
                optimizer,
                step_size=0.003,
                epsilon=0.031,
                perturb_steps=10,
                beta=1.0,
                distance='l_inf'):
    # define KL-loss
    criterion_kl = nn.KLDivLoss(size_average=False)
    model.eval()
    batch_size = len(x_natural)
    # generate adversarial example
    x_adv = x_natural.detach() + 0.001 * torch.randn(x_natural.shape).cuda().detach()
    if distance == 'l_inf':
        for _ in range(perturb_steps):
            x_adv.requires_grad_()
            with torch.enable_grad():
                loss_kl = criterion_kl(F.log_softmax(model(x_adv), dim=1),
                                       F.softmax(model(x_natural), dim=1))
            grad = torch.autograd.grad(loss_kl, [x_adv])[0]
            x_adv = x_adv.detach() + step_size * torch.sign(grad.detach())
            x_adv = torch.min(torch.max(x_adv, x_natural - epsilon), x_natural + epsilon)
            x_adv = torch.clamp(x_adv, 0.0, 1.0)
    elif distance == 'l_2':
        delta = 0.001 * torch.randn(x_natural.shape).cuda().detach()
        delta = Variable(delta.data, requires_grad=True)

        # Setup optimizers
        optimizer_delta = optim.SGD([delta], lr=epsilon / perturb_steps * 2)

        for _ in range(perturb_steps):
            adv = x_natural + delta

            # optimize
            optimizer_delta.zero_grad()
            with torch.enable_grad():
                loss = (-1) * criterion_kl(F.log_softmax(model(adv), dim=1),
                                           F.softmax(model(x_natural), dim=1))
            loss.backward()
            # renorming gradient
            grad_norms = delta.grad.view(batch_size, -1).norm(p=2, dim=1)
            delta.grad.div_(grad_norms.view(-1, 1, 1, 1))
            # avoid nan or inf if gradient is 0
            if (grad_norms == 0).any():
                delta.grad[grad_norms == 0] = torch.randn_like(delta.grad[grad_norms == 0])
            optimizer_delta.step()

            # projection
            delta.data.add_(x_natural)
            delta.data.clamp_(0, 1).sub_(x_natural)
            delta.data.renorm_(p=2, dim=0, maxnorm=epsilon)
        x_adv = Variable(x_natural + delta, requires_grad=False)
    else:
        x_adv = torch.clamp(x_adv, 0.0, 1.0)
    model.train()

    x_adv = Variable(torch.clamp(x_adv, 0.0, 1.0), requires_grad=False)
    # zero gradient
    optimizer.zero_grad()
    # calculate robust loss
    logits = model(x_natural)
    loss_natural = F.cross_entropy(logits, y)
    loss_robust = (1.0 / batch_size) * criterion_kl(F.log_softmax(model(x_adv), dim=1),
                                                    F.softmax(model(x_natural), dim=1))
    loss = loss_natural + beta * loss_robust
    return loss

def trades_train(model, device, train_loader, optimizer, epoch, train_config):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)

        optimizer.zero_grad()

        # calculate robust loss
        loss = trades_loss(model=model,
                           x_natural=data,
                           y=target,
                           optimizer=optimizer,
                           step_size=train_config['step_size'],
                           epsilon=train_config['epsilon'],
                           perturb_steps=train_config['num_steps'],
                           beta=6.0)
        loss.backward()
        optimizer.step()

        # print progress
        if batch_idx % 100 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                       100. * batch_idx / len(train_loader), loss.item()))

def adjust_learning_rate(optimizer, epoch):
    """decrease the learning rate"""
    lr = 0.1
    if epoch >= 75:
        lr = 0.1 * 0.1
    if epoch >= 90:
        lr = 0.1 * 0.01
    if epoch >= 100:
        lr = 0.1 * 0.001
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr

def _pgd_whitebox(model,
                  X,
                  y,
                  device,
                  epsilon=0.031,
                  num_steps=20,
                  step_size=2. / 255,):
    out = model(X)
    err = (out.data.max(1)[1] != y.data).float().sum()
    X_pgd = Variable(X.data, requires_grad=True)
    random = True
    if random:
        random_noise = torch.FloatTensor(*X_pgd.shape).uniform_(-epsilon, epsilon).to(device)
        X_pgd = Variable(X_pgd.data + random_noise, requires_grad=True)

    for _ in range(num_steps):
        opt = optim.SGD([X_pgd], lr=1e-3)
        opt.zero_grad()

        with torch.enable_grad():
            loss = nn.CrossEntropyLoss()(model(X_pgd), y)
        loss.backward()
        eta = step_size * X_pgd.grad.data.sign()
        X_pgd = Variable(X_pgd.data + eta, requires_grad=True)
        eta = torch.clamp(X_pgd.data - X.data, -epsilon, epsilon)
        X_pgd = Variable(X.data + eta, requires_grad=True)
        X_pgd = Variable(torch.clamp(X_pgd, 0, 1.0), requires_grad=True)
    err_pgd = (model(X_pgd).data.max(1)[1] != y.data).float().sum()
    print('err pgd (white-box): ', err_pgd)
    return err, err_pgd

def eval_adv_test_whitebox(model, device, data, target, test_config):
    """
    evaluate model by white-box attack
    """
    model.eval()
    robust_err_total = 0
    natural_err_total = 0

    data, target = data.to(device), target.to(device)
    # pgd attack
    X, y = Variable(data, requires_grad=True), Variable(target)
    err_natural, err_robust = _pgd_whitebox(model, X, y, device = device, \
        epsilon=test_config['epsilon'], num_steps=test_config['num_steps'], step_size=test_config['step_size'])
    robust_err_total += err_robust
    natural_err_total += err_natural
    print('natural_err_total: ', natural_err_total)
    print('robust_err_total: ', robust_err_total)
    acc = 1 - 1. * natural_err_total / len(data)
    rob = 1 - 1. * robust_err_total / len(data)
    return acc, rob

def pgd(model, data, target, epsilon, step_size, num_steps,loss_fn,category,rand_init):
    model.eval()
    if category == "trades":
        x_adv = data.detach() + 0.001 * torch.randn(data.shape).cuda().detach() if rand_init else data.detach()
    if category == "Madry":
        x_adv = data.detach() + torch.from_numpy(np.random.uniform(-epsilon, epsilon, data.shape)).float().cuda() if rand_init else data.detach()
        x_adv = torch.clamp(x_adv, 0.0, 1.0)
    for k in range(num_steps):
        x_adv.requires_grad_()
        output = model(x_adv)
        model.zero_grad()
        with torch.enable_grad():
            if loss_fn == "cent":
                loss_adv = nn.CrossEntropyLoss(reduction="mean")(output, target)
        loss_adv.backward()
        eta = step_size * x_adv.grad.sign()
        x_adv = x_adv.detach() + eta
        x_adv = torch.min(torch.max(x_adv, data - epsilon), data + epsilon)
        x_adv = torch.clamp(x_adv, 0.0, 1.0)
    return x_adv

def eval_clean(model, data, target):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        # for data, target in test_loader:
        data, target = data.cuda(), target.cuda()
        output = model(data)
        test_loss += nn.CrossEntropyLoss(reduction='mean')(output, target).item()
        pred = output.max(1, keepdim=True)[1]
        correct += pred.eq(target.view_as(pred)).sum().item()
    test_loss /= len(data)
    log = 'Natrual Test Result ==> Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)'.format(
        test_loss, correct, len(data),
        100. * correct / len(data))
    # print(log)
    test_accuracy = correct / len(data)
    return test_loss, test_accuracy

def eval_robust(model, data, target, perturb_steps=1, epsilon=0.031, step_size=0.031,loss_fn="cent", category="Madry",rand_init=False):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.enable_grad():
        # for data, target in test_loader:
        data, target = data.cuda(), target.cuda()
        x_adv = pgd(model,data,target,epsilon,step_size,perturb_steps,loss_fn,category,rand_init=rand_init)
        output = model(x_adv)
        test_loss += nn.CrossEntropyLoss(reduction='mean')(output, target).item()
        pred = output.max(1, keepdim=True)[1]
        correct += pred.eq(target.view_as(pred)).sum().item()
    test_loss /= len(data)
    log = 'Attack Setting ==> Loss_fn:{}, Perturb steps:{}, Epsilon:{}, Step dize:{} \n Test Result ==> Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)'.format(loss_fn,perturb_steps,epsilon,step_size,
        test_loss, correct, len(data),
        100. * correct / len(data))
    # print(log)
    test_accuracy = correct / len(data)
    return test_loss, test_accuracy

