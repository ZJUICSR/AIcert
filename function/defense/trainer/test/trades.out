nohup: ignoring input
got GPU
cuda:0
Step 1: Load the MNIST dataset
Files already downloaded and verified
Files already downloaded and verified
Step 2: Create the model
Step 2a: Define the optimizer
Step 4: Train the TRADES classifier
/mnt/data/miniconda3/envs/art/lib/python3.9/site-packages/torch/nn/_reduction.py:44: UserWarning: size_average and reduce args will be deprecated, please use reduction='sum' instead.
  warnings.warn(warning.format(ret))
Train Epoch: 1 [0/50000 (0%)]	Loss: 2.443996
Train Epoch: 1 [6400/50000 (13%)]	Loss: 2.276883
Train Epoch: 1 [12800/50000 (26%)]	Loss: 2.110475
Train Epoch: 1 [19200/50000 (38%)]	Loss: 2.163791
Train Epoch: 1 [25600/50000 (51%)]	Loss: 2.112572
Train Epoch: 1 [32000/50000 (64%)]	Loss: 2.093357
Train Epoch: 1 [38400/50000 (77%)]	Loss: 2.079063
Train Epoch: 1 [44800/50000 (90%)]	Loss: 2.068716
Step 5: Evaluate the ART classifier on benign test examples
err pgd (white-box):  tensor(77., device='cuda:0')
natural_err_total:  tensor(65., device='cuda:0')
robust_err_total:  tensor(77., device='cuda:0')
Accuracy on benign test examples: 35.000003814697266%
Step 7: Evaluate the ART classifier on adversarial test examples
Accuracy on adversarial test examples: 23.000001907348633%
detect rate:  tensor(0.2300, device='cuda:0')
